sql: fix the rewrite for logic tests

knz opened the pull request:
Reverts #27970.
Fixes #28025.

The change was incorrect with regards to queries where the amount of
whitespace per column is sensitive, for example the output of EXPLAIN.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28027)
<!-- Reviewable:end -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28027)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/logictest/logic.go, line 1897 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28027#-LIfKm4N3-ggPIJ2AqOQ:-LIfKm4N3-ggPIJ2AqOR:bz0nnra) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ab7e899617492229a241bfdf89ba7ed8e46a117/pkg/sql/logictest/logic.go#L1897)):*
> ```Go
> 			} else {
> 				// Emit the actual results.
> 				// We cannot use formatValues() here because the query rows
> ```

`formatValues` doesn't remove whitespace (except end-of-line trailing whitespace), why can't we use it with `actualResultsRaw`?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28027)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/logictest/logic.go, line 1897 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28027#-LIfKm4N3-ggPIJ2AqOQ:-LIfRIfg72XU3HLlAPxn:b62zq5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ab7e899617492229a241bfdf89ba7ed8e46a117/pkg/sql/logictest/logic.go#L1897)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`formatValues` doesn't remove whitespace (except end-of-line trailing whitespace), why can't we use it with `actualResultsRaw`?
</blockquote></details>

Because `actualResultsRaw` contains the space-delimited *words* from the query, not the result rows. So with N output columns, `formatValues` will cause the first N words to be emitted on the first line, then ext N words on the 2nd line, etc. If a line contained multiple words per column, that messes up the output.

(try it!)


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I did try it..  I believe `actualResults` has the space-delimited words, as opposed to `actualResultsRaw` which AFAICT always has N values if there are N columns.

knz commented:
ok then something else is going on my branch. I will investigate further.

andy-kimball reviewed (commented):
What was the reason we want to output rows in sorted order to begin with? The PR that made this change does not explain the motivation. +@mjibson, can you explain what motivated you to make that change?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28027#-:-LIfXVco6JjGle1fML0i:btfjifs)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
Also, a word of caution: I learned the hard way that any time you change this code, you need to always run `--rewrite-results-in-testfiles` to make sure you didn't break it, *and* you need to run the SQL Lite `-bigtest` suite (both with and without the `rewrite` option). It's very easy to fix one thing and inadvertently break another thing.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28027#-:-LIfYb8a3ActR7h7qzTn:bp25ka4)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz commented:
Andy thanks for your warning. I am careful to test with `-rewrite-results` but I hadn't thought to confirm with `-bigtest`. Will do.

RaduBerinde commented:
Note that `--rewrite-results-in-testfiles` only does something when the results don't match (would have been a test failure). So that code can only be tested by modifying a testcase output first.

knz commented:
Yes I have noticed. To test it I always delete the reference output first.

andy-kimball reviewed (commented):
Sounds good. I also remembered another thing I did when testing: I temporarily changed the `--rewrite-results-in-testfiles` to always rewrite (so I didn't need to delete the reference output), and then checked that rewriting *all* files only differed in inconsequential whitespace. One interesting example of a diff is in `builtin_function`, line 1639:
```
SELECT to_uuid('63616665-6630-3064-6465-616462656566'),
       to_uuid('{63616665-6630-3064-6465-616462656566}'),
       to_uuid('urn:uuid:63616665-6630-3064-6465-616462656566'),
       from_uuid(b'cafef00ddeadbeef'),
       to_uuid(from_uuid(b'cafef00ddeadbeef')),
       from_uuid(to_uuid('63616665-6630-3064-6465-616462656566'))
```
Whoever wrote this case seems to have deliberately put each output column on its own separate line, rather than all on one line, like `--rewrite-results-in-testfiles` would do. The `logic_test` accepts either variation of expected results.

Another test that has tricky spacing issues is `logic_test/computed`, since `SHOW CREATE TABLE` has multiple columns, with one of the columns having newlines in its value.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28027#-:-LIfd6NQ9-85LEa7Q66x:b-k2xerd)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz committed:
revert "sql/logictest: when rewriting, emit sorted results"

Reverts #27970.

The change was incorrect with regards to queries where the amount of
whitespace per column is sensitive, for example the output of EXPLAIN.

Release note: None

knz commented:
Radu you were right all along, reverting #27970 was all it takes.

Sorry for the inconvenience.

knz commented:
(I changed the PR, RFAL)

RaduBerinde reviewed (approved):
LGTM. Thanks for the fix!

knz commented:
bors r+

RaduBerinde commented:
Did you get a chance to run the bigtest? I see the following failure locally:
```
$ make test PKG=./pkg/sql/logictest/ TESTS=TestLogic TESTFLAGS='-config local-opt -d /go/src/github.com/cockroachdb/sqllogictest/test/select4.test -show-logs -show-sql'

/go/src/github.com/cockroachdb/sqllogictest/test/select4.test:3091: expected 768 results, but found 1152
```

The rows look like `996  297  177  577  602  table tn1 row 128`. There are 128 rows so the white space isn't being handled correctly (each word counts as a different result, leading to 9*128 instead of 6*128).

RaduBerinde commented:
Hm, scratch that. I didn't pass the `-bigtest` flag which changes the way the results are handled.

mjibson commented:
My motivation from the original PR was that for queries that have `rowsort` in them and aren't meaningfully sorted by an ORDER BY, I want the rewritten results to be the same each time instead of possibly change on every rewrite based on the non-deterministic order they were returned by the server. The specific test I was trying to work with happened to not meet this criteria because I misunderstood what `ORDER BY 1` did (I thought it was a no-op) which I think led to this being a not-so-much working PR.
